src/Avalonia.Controls/Documents/TextElement.cs

```diff
@@ -0,0 +1,145 @@
+
+        /// <summary>
+        /// Defines the <see cref="Foreground"/> property.
+        /// </summary>
+        public static readonly AttachedProperty<IBrush> ForegroundProperty =
+            TextBlock.ForegroundProperty.AddOwner<Inline>();
+
+        static TextElement()
+        {
+            InvalidatesTextElement<TextElement>(
+                BackgroundProperty,
+                FontFamilyProperty,
+                FontSizeProperty,
+                FontStyleProperty,
+                FontWeightProperty,
+                ForegroundProperty);
+        }
+        
+        /// <summary>
+        /// Gets or sets a brush used to paint the control's background.
+        /// </summary>
+        public IBrush Background
+        {
+            get { return GetValue(BackgroundProperty); }
+            set { SetValue(BackgroundProperty, value); }
+        }
+
+        /// <summary>
+        /// Gets or sets the font family.
+        /// </summary>
+        public FontFamily FontFamily
+        {
+            get { return GetValue(FontFamilyProperty); }
+            set { SetValue(FontFamilyProperty, value); }
+        }
+
+        /// <summary>
+        /// Gets or sets the font size.
+        /// </summary>
+        public double FontSize
+        {
+            get { return GetValue(FontSizeProperty); }
+            set { SetValue(FontSizeProperty, value); }
+        }
+
+        /// <summary>
+        /// Gets or sets the font style.
+        /// </summary>
+        public FontStyle FontStyle
+        {
+            get { return GetValue(FontStyleProperty); }
+            set { SetValue(FontStyleProperty, value); }
+        }
+
+        /// <summary>
+        /// Gets or sets the font weight.
+        /// </summary>
+        public FontWeight FontWeight
+        {
+            get { return GetValue(FontWeightProperty); }
+            set { SetValue(FontWeightProperty, value); }
+        }
+
+        /// <summary>
+        /// Gets or sets a brush used to paint the text.
+        /// </summary>
+        public IBrush Foreground
+        {
+            get { return GetValue(ForegroundProperty); }
+            set { SetValue(ForegroundProperty, value); }
+        }
+        
+        /// <summary>
+        /// Raised when the visual representation of the text element changes.
+        /// </summary>
+        public event EventHandler Invalidated;
+        
+        /// <summary>
+        /// Marks properties as invalidating the text element's visual representation.
+        /// </summary>
+        /// <param name="properties">The properties.</param>
+        /// <remarks>
+        /// After a call to this method in a text element's static constructor, any change to a
+        /// property in the <paramref name="properties"/> collection will cause
+        /// <see cref="Invalidate"/> to be called on the element.
+        /// </remarks>
+        protected static void InvalidatesTextElement<T>(params AvaloniaProperty[] properties)
```

Comment:
Instead of this, to avoid delegate allocation, you can override OnPropertyChanged